src/services/soft-crud-service.ts

```diff
@@ -0,0 +1,93 @@
+import {
+  AndClause,
+  Condition,
+  DataObject,
+  Filter,
+  OrClause,
+} from '@loopback/repository';
+import {SoftDeleteEntity} from '../models';
+
+export class SoftCrudService {
+  /**
+   * Ensures `{ deleted: false }` condition to be in filter.where
+   * @param filter filter reference object that needs be altered to include `deleted: false` condition
+   */
+  static modifyWhereFilter<E extends SoftDeleteEntity>(
+    filter: Filter<E>,
+    conditionToEnsure?: Condition<E>,
+  ) {
+    filter.where = filter.where ?? {};
+
+    conditionToEnsure = conditionToEnsure ?? ({deleted: false} as Condition<E>);
+
+    const hasAndClause = (filter.where as AndClause<E>).and?.length > 0;
+    const hasOrClause = (filter.where as OrClause<E>).or?.length > 0;
```

Comment:
It could have both na?

Reply:
Actually no, the where filter can have only one root key either "and" or "or", they can surely have nested like below

```json
{
  "and": [
    { "key1": "--", "andKey2": "--" },
    {
      "or": [ ... ]
    }
  ]
}
```


From the types
```ts
export type Where<MT extends object = AnyObject> = Condition<MT> | AndClause<MT> | OrClause<MT>;
```

Reply:
Can't trust Loopback4 types 🙂 , just try this payload in a request for a model ->

```json
{
  "where": {
    "or": [{"id": 1}, {"id": 2}],
    "and": [{"name": {"ilike": "oe"}}]
  }
}
````

In this only the `or` clause would work, but If I use ->

```json
{
  "where": {
    "and": [{"id": 1}, {"id": 2}],
    "or": [{"name": {"ilike": "oe"}}]
  }
}
````

only the and `and` clause would work. This is something I've tried for postgres connector but could be true for other connectors as well.

Reply:
Oh, then the current change is fair based on this, right?

Because even the [docs](https://loopback.io/doc/en/lb4/Where-filter#and-and-or-operators) hints about using one of them as the base level key:
<img width="883" alt="image" src="https://user-images.githubusercontent.com/110156023/221369478-c1dd2299-4760-44bd-9cbe-3a000e0a3e2d.png">

So If someone uses both and expects soft delete to work correctly he wouldn't get that since it's logically wrong.

Reply:
Nope, the current (or the previous code)  would give wrong result as only the `and` clause would be modified but only the `or` clause would actually run.

>Because even the [docs](https://loopback.io/doc/en/lb4/Where-filter#and-and-or-operators) hints about using one of them as the base level key:

Either the docs are wrong or the implementation is wrong, because of as of now, there is nothing stopping the user from sending both operators at the same level and api does work with them if you try. If both are not allowed at base level, the filter should either be considered invalid or should have a well defined behaviour. Neither of which is mentioned in the docs. 

>So If someone uses both and expects soft delete to work correctly he wouldn't get that since it's logically wrong.

Why is it logically wrong to use `and` and  `or` at same level?

Reply:
> Why is it logically wrong to use and and or at same level?

Because the framework has to guess the query part between `and` and `or` conditions which user can't explicitly define this way. (so not logically wrong I feel it's more declarative when using just one base level key.)

But I got your point it's just the matter of removing `else if` to `if` for the `hasOrClause` condition.

Reply:
>Because the framework has to guess the query part between and and or conditions which user can't explicitly define this way. 

Why it has to guess? for direct field queries, it assumes an `and` operator, ideally it should have been same for these two clauses as well - 
```
{
  "field1": "value1",
  "field2": "value2"
 }
```

Reply:
Fixed.

PS: Gonna raise this to loopback as well. :)